datasource/hcp-packer-image/data.go

```diff
@@ -0,0 +1,153 @@
+	// "aws", "azure", or "gce".
+	CloudProvider string `mapstructure:"cloud_provider"`
+	// The specific Packer builder or post-processor used to create the image.
+	ComponentType string `mapstructure:"component_type"`
+	// The date and time at which the image was created.
+	CreatedAt string `mapstructure:"created_at"`
+	// The id of the build that created the image. This is a ULID, which is a
+	// unique identifier similar to a UUID. It is created by the HCP Packer
+	// Registry when an build is first created, and is unique to this build.
+	BuildID string `mapstructure:"build_id"`
+	// The iteration id. This is a ULID, which is a unique identifier similar
+	// to a UUID. It is created by the HCP Packer Registry when an iteration is
+	// first created, and is unique to this iteration.
+	IterationID string `mapstructure:"iteration_id"`
+	// The UUID associated with the Packer run that created this image.
+	PackerRunUUID string `mapstructure:"packer_run_uuid"`
+	// The iteration id. This is a ULID, which is a unique identifier similar
+	// to a UUID. It is created by the HCP Packer Registry when an iteration is
+	// first created, and is unique to this iteration.
+	ID string `mapstructure:"id"`
+	// ID or URL of the remote cloud image as given by a build.
+	ImageID string `mapstructure:"image_id"`
+	// The cloud region as given by `packer build`. eg. "ap-east-1".
+	// For locally managed clouds, this may map instead to a cluster, server
+	// or datastore.
+	Region string `mapstructure:"region"`
+}
+
+func (d *Datasource) OutputSpec() hcldec.ObjectSpec {
+	return (&DatasourceOutput{}).FlatMapstructure().HCL2Spec()
+}
+
+func (d *Datasource) Execute() (cty.Value, error) {
+	cli, err := packerregistry.NewClient()
+	if err != nil {
+		return cty.NullVal(cty.EmptyObject), err
+	}
+
+	// Load channel.
+	log.Printf("[INFO] Reading info from HCP Packer registry (%s) [project_id=%s, organization_id=%s, iteration_id=%s]",
+		d.config.Bucket, cli.ProjectID, cli.OrganizationID, d.config.IterationID)
+
+	iteration, err := packerregistry.GetIterationFromId(context.TODO(), cli, d.config.Bucket, d.config.IterationID)
+	if err != nil {
+		return cty.NullVal(cty.EmptyObject), fmt.Errorf("error retrieving "+
+			"image iteration from HCP Packer registry: %s", err.Error())
+	}
+
+	output := DatasourceOutput{}
+
+	for _, build := range iteration.Builds {
+		if build.CloudProvider == d.config.CloudProvider {
+			for _, image := range build.Images {
+				if image.Region == d.config.Region {
+					// This is the desired image.
+					output = DatasourceOutput{
+						CloudProvider: build.CloudProvider,
+						ComponentType: build.ComponentType,
+						CreatedAt:     image.CreatedAt.String(),
+						BuildID:       build.ID,
+						IterationID:   build.IterationID,
+						PackerRunUUID: build.PackerRunUUID,
+						ID:            image.ID,
+						ImageID:       image.ImageID,
+						Region:        image.Region,
+					}
+				}
+			}
+		}
```

Comment:
nit: Use guard  clause over the nested if statements.
```go
	if build.CloudProvider != d.config.CloudProvider {
		continue
	}
	for _, image := range build.Images {
		if image.Region != d.config.Region {
			continue
		}
		// This is the desired image.
		output = DatasourceOutput{
			CloudProvider: build.CloudProvider,
			ComponentType: build.ComponentType,
			CreatedAt:     image.CreatedAt.String(),
			BuildID:       build.ID,
			IterationID:   build.IterationID,
			PackerRunUUID: build.PackerRunUUID,
			ID:            image.ID,
			ImageID:       image.ImageID,
			Region:        image.Region,
		}
	}
```

Reply:
Sure!